### PR TITLE
Fix for Tab Completion in Nested mode

### PR DESCRIPTION
**Pull request by davidBar-On**
# PR Summary

Suggested fix to issue #3810 for Tab Completion in Nested mode.  This is to make sure all local variables and aliases are included in the tab-completion or Get-Alias.  The change is to add local variable and aliases to `CommandScope` before executing read line.  The changes are required at least when entering the nested mode from a script by Write-Debug suspend selection.  Without this change, all variables and aliases defined in the script or after the suspend are **not** included in the tab-completion or Get-Alias.

 Although the suggested change seem to work well, there are some potential issues with the it.  However, it seem that if required they can be better resolved when the suggested changes are available (this is why I submit this PR without trying to resolve these issues first).

**The potential issues are:**
1. **No test cases were added**.  I didn't find any available test of the **Write-Debug Suspend mode**, so I assume this mode is tested only manually.
2. **There may be a better place to add the changes**. It may be that that adding the local variables/aliases to the `CommandScope` should be done when it is generated (in case it is re-recreated before each read line, to make sure latest locals are added).
3. **New API may be defined for adding the locals to the scope**.   However, it seemed to me that until there is a decision about final code for adding the variables to the scope, it is better to have all the code in one place, so it will be clear what is done.

## PR Context

Resolve issue #3810.

## PR Checklist

- [X ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [X] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [X] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [X] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

**Comment by PaulHigin** [comment]
Leaking variables across scopes does not seem like the right way to fix this rare and unusual scenario.
@daxian-dbw can you take a look?

**Comment by adityapatwardhan** [comment]
I would wait on review from @daxian-dbw who is on vacation right now. Maybe @PowerShell/wg-powershell-engine can help in reviewing this.

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 1]
```diff
@@ -358,6 +359,52 @@ internal void SetCurrentScopeToExecutionScope()
...
+            // so they be accecible by the command.
+            // Required in nested shell for tab completion and Get-Aliases (issue #3810).
+            // Note: "TabExpansion2" is included since it seems it is being used for read line
+            //       when PS "Cannot load PSReadline module.  Console is running without PSReadline". 
+            if (_command.CommandInfo.Name == "PSConsoleHostReadLine" || _command.CommandInfo.Name == "TabExpansion2")
+            {
+                SessionStateScope localScope = _previousCommandSessionState.CurrentScope;
+                if (localScope != null)
+                {
+                    if (localScope != CommandSessionState.CurrentScope)
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs:372]: Easier to condense this to:

```csharp
if (localScope is not null && localScope != CommandSessionState.CurrentScope)
{
...
}
```

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 2]
```diff
@@ -358,6 +359,52 @@ internal void SetCurrentScopeToExecutionScope()
...
+                    {
+                        // Add local variables from previous scope.
+                        foreach (KeyValuePair<string, PSVariable> entry in localScope.Variables)
+                        {
+                            // No need to re-include all-scope or non simple mutable variables...
+                            if (!entry.Value.IsAllScope && entry.Value.Options == ScopedItemOptions.None && entry.Value is not NullVariable)
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs:378]:
```suggestion
                            if (!entry.Value.IsAllScope
                                && entry.Value.Options == ScopedItemOptions.None
                                && entry.Value != NullVariable)
```

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 3]
```diff
@@ -358,6 +359,52 @@ internal void SetCurrentScopeToExecutionScope()
...
+                            {
+                                CommandSessionState.CurrentScope.SetVariable(entry.Key, entry.Value, false, false, _previousCommandSessionState);
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs:380]: Please label boolean parameters

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 4]
```diff
@@ -358,6 +359,52 @@ internal void SetCurrentScopeToExecutionScope()
...
+                            CommandSessionState.CurrentScope.DottedScopes.Push(localScope.LocalsTuple);
+                        }
+
+                        // Add aliases from previous scope.
+                        if (localScope.AliasTable != null)
+                        {
+                            foreach (AliasInfo entry in localScope.AliasTable)
+                            {
+                                // Make sure the alias isn't private.
+                                if ((entry.Options & ScopedItemOptions.Private) == 0)
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs:397]:
```suggestion
                                if (!entry.Option.HasFlag(ScopedItemOptions.Private))
```

File: src/System.Management.Automation/engine/CommandProcessorBase.cs [hunk 5]
```diff
@@ -358,6 +359,52 @@ internal void SetCurrentScopeToExecutionScope()
...
+                                {
+                                    // `force` parameter is set to `true` since otherise SetAliasItem throws an exception.
+                                    CommandSessionState.CurrentScope.SetAliasItem(entry, true);
```
Comment (rjmholt) [inline src/System.Management.Automation/engine/CommandProcessorBase.cs:400]: Label boolean parameter

**Comment by vexx32** [comment]
We looked over this in the Engine working group today (8/5/2021). We're in agreement with @PaulHigin's assertion here; copying variables to additional scopes like this may have unintended side effects. Especially as the condition we're using is based around the command name, it may be possible possible to trigger the behaviour in cases that we aren't expecting by simply defining new functions matching those names.

From the discussion in the linked issue and the code change here, it seems like a potentially more appropriate solution would be to either:

- Give the tab completion code access to the previous scope in these specific cases, _or_
- Ensure that the scope given to the ReadLine call is a child of the original, currently-executing scope.

This change seems somewhat heavy-handed; we're not just copying variable names, we're also copying values. If there's a more direct solution available, it would be preferable to go that route instead.

**Pull request closed without merging**